Feat: improve authentication process

niccoloMG opened the pull request:
## Description

Improve our authentication process by 

- introducing a navbar with sign out
- linking login and signup forms
- adding a forgot password form
- displaying success and error alerts after submissions

## Task

[Notion Task Link](https://www.notion.so/Tasks-538c1c6dbb2f4cba9a11c204ff418786?p=69424ad72d8b4de7bd7e673b8e2ab102&pm=s)

## Screenshots (optional)

Login:
![image](https://github.com/berk245/magic-castle/assets/52704542/ea36fb69-2a96-41ca-9dd1-ce415a7ab202)

Reset Password:
![image](https://github.com/berk245/magic-castle/assets/52704542/277cc22a-f242-45a6-8daf-70d24d0138e9)

Error message:
![image](https://github.com/berk245/magic-castle/assets/52704542/f00bb010-424d-46c4-9da2-66391b7cc79a)

Success Message:
![image](https://github.com/berk245/magic-castle/assets/52704542/21b3ec3b-97e1-4a66-9eb2-70d462ed0794)

niccoloMG committed:
Create a navbar component

niccoloMG committed:
Add user prompt to navbar

niccoloMG committed:
Add a simple navbar

niccoloMG committed:
Render navbar in main layout

niccoloMG committed:
Add links between login and signup forms

niccoloMG committed:
Add reset password component

niccoloMG committed:
Add a link to reset password form from login

niccoloMG committed:
Add a reset password auth handler

niccoloMG committed:
Trigger reset password function on form submit

niccoloMG committed:
Format auth.js file

niccoloMG committed:
Add an alert component with danger and success types

niccoloMG committed:
Improve code formatting in login view

niccoloMG committed:
Render alerts after submission success or fail in reset password

niccoloMG committed:
Render alerts after submission success or fail after signup

berk245 commented on src/lib/components/Alert.svelte:
```diff
@@ -0,0 +1,86 @@
+	{#if alertType === 'success'}
+		<div
+			id="alert-additional-content-3"
+			class="p-4 mb-4 text-green-800 border border-green-300 rounded-lg bg-green-50 dark:bg-gray-800 dark:text-green-400 dark:border-green-800"
```
It looks like we are creating the same containers here, also similar alert components with different styles. Let's try to not repeat ourselves. I'd suggest creating one alert component and dynamically assigning classes to it based on its type (success or danger for now)

If that is too complicated we can create separate components for success and danger alerts for now.

tonyjojo replied:
The DRY principle is very important, I agree
I think they should be just one component that accepcts props to change the layout and copy

niccoloMG replied:
Okay I will try to dynamically set the classes based on variant prop

niccoloMG replied:
Added a function to dynamically set colors and class strings. I used a switch statement in case we want to add new alert types in the future.

berk245 commented on src/lib/handlers/auth.js:
```diff
@@ -8,6 +13,10 @@ export const authHandlers = {
 		await signInWithEmailAndPassword(auth, email, password);
 	},
+	resetPassword: async ({ email }) => {
+		let serverResponse = await sendPasswordResetEmail(auth, email);
```
Do we need to return the server response here? Why is it different than the other functions?

niccoloMG replied:
It was for testing something. Thanks for pointing it out

niccoloMG replied:
Fixed!

berk245 commented on src/routes/(auth)/reset-password/+page.svelte:
```diff
@@ -0,0 +1,72 @@
+				>
+					{#if !$authStore.loading}
+						Reset Password
+					{:else}
```
I am not sure if the loading feedback is working as intended. Can you make sure it's working and implement it for the other forms as well?

niccoloMG replied:
I will add the loading feedback in a separate PR.

berk245 reviewed (changes requested):
Good additions.

Let's iron out a few things mentioned in the comments and merge.

You can probably extract multiple smaller, easier to review PRs from this one. Something to keep in mind for the next time1

niccoloMG committed:
Add classes object to dynamically assign classes based on alert type

niccoloMG committed:
Remove unused import

niccoloMG committed:
Improve dynamic class assignment process in alert

Add a getAlertClasses function to dynamically create class strings
instead of hardcoding them in an object.

niccoloMG committed:
Remove unnecessary variable in auth handlers

berk245 reviewed (approved):
Great work!

berk245 commented on src/lib/components/Alert.svelte:
```diff
@@ -0,0 +1,64 @@
+			{alertMessage}
+		</div>
+		<div class="flex">
+			<a href={backToRoute}>
```
This doesn't work for signup error because the go back route is the same as the current route. Can you please fix that?

niccoloMG replied:
Fixed it by adding target='_self'

niccoloMG committed:
Fix no redirection on signup error alert by adding a self target

berk245 reviewed (approved):
Great work!

tonyjojo reviewed (approved):
Good job tackling the changes there Nicco
